File: problem_3/haskell/problem3.hs
module Main where

import Control.Parallel.Strategies

factors count top_number
  | count >= top_number = []
  | zero_mod count = count : new_top: factors (count + 2) new_top
  | otherwise = factors (count + 2) top_number
  where new_top = top_number `div` count

zero_mod divisor
  | divide == 0 = True 
  | otherwise = False
  where divide = mod 600851475143 divisor

--prime :: Integer -> Integral -> Bool
prime divisor divided
  | divided == 1 = True
  | divisor >= sqrt_divided = True
  | mod divided divisor == 0 = False
  | otherwise = next_prime
  where next_prime = prime (divisor + 2) divided
        sqrt_divided = round . sqrt . fromIntegral $ divided

is_prime :: Integer -> Bool
is_prime input = prime 3 input

main :: IO ()
main = print . snd . maximum $ zip y x
  where x = factors 3 600851475143
        y = parMap rwhnf  is_prime x

File: problem_3/perl/problem3.pl
#!/usr/bin/perl

use strict;
use warnings;

my $top_number = 600851475143;

sub is_prime
{
  my ($divided) = @_;
  return 1 if ($divided == 1);

  my $divisor = 3;
  my $sqrt_divided = sqrt($divided);

  while($divisor <= $sqrt_divided)
  {
    return 0 unless ($divided % $divisor);
    $divisor += 2;
  }

  1;
}

my @f;
my $new_top = $top_number;

for(my $i = 3; $i <= $new_top; $i += 2)
{
  unless ($top_number % $i)
  {
    if (is_prime($i))
    {
      push @f, $i;
    }
  }
    $new_top = $top_number / $i;
}

print $f[-1];
